nft_ingester/src/backfiller.rs

```diff
@@ -34,6 +36,145 @@ pub const GET_DATA_FROM_BG_RETRIES: u32 = 5;
+    rx: Receiver<()>,
+) -> Result<(), JoinError>
+where
+    C: BlockConsumer,
+    P: BlockProducer,
+    S: SlotGetter,
+{
+    info!("Running slot force persister...");
+
+    force_reingestable_transactions_parser
+        .parse_transactions(rx)
+        .await;
+
+    info!("Force slot persister finished working.");
+
+    Ok(())
+}
+
+pub enum BackfillSource {
+    Bigtable(Arc<BigTableClient>),
+    Rpc(Arc<BackfillRPC>),
+}
+
+impl BackfillSource {
+    pub async fn new(
+        ingester_config: &IngesterConfig,
+        backfiller_config: &BackfillerConfig,
+    ) -> Self {
+        match ingester_config.backfiller_source_mode {
+            BackfillerSourceMode::Bigtable => Self::Bigtable(Arc::new(
+                connect_new_bigtable_from_config(backfiller_config.clone())
+                    .await
+                    .unwrap(),
+            )),
+            BackfillerSourceMode::RPC => Self::Rpc(Arc::new(BackfillRPC::connect(
+                ingester_config.backfill_rpc_address.clone(),
+            ))),
+        }
+    }
+}
+
+#[async_trait]
+impl SlotsGetter for BackfillSource {
+    async fn get_slots(
+        &self,
+        collected_key: &Pubkey,
+        start_at: u64,
+        rows_limit: i64,
+    ) -> Result<Vec<u64>, UsecaseError> {
+        match self {
+            BackfillSource::Bigtable(bigtable) => {
+                bigtable
+                    .big_table_inner_client
+                    .get_slots(collected_key, start_at, rows_limit)
+                    .await
+            }
+            BackfillSource::Rpc(rpc) => rpc.get_slots(collected_key, start_at, rows_limit).await,
+        }
+    }
+}
+
+#[async_trait]
+impl BlockProducer for BackfillSource {
+    async fn get_block(
+        &self,
+        slot: u64,
+        backup_provider: Option<Arc<impl BlockProducer>>,
+    ) -> Result<UiConfirmedBlock, StorageError> {
+        match self {
+            BackfillSource::Bigtable(bigtable) => bigtable.get_block(slot, backup_provider).await,
+            BackfillSource::Rpc(rpc) => rpc.get_block(slot, backup_provider).await,
+        }
+    }
+}
+
+pub async fn run_perpetual_slot_collection(
+    backfiller_clone: Arc<Backfiller<BackfillSource>>,
+    rpc_backfiller_clone: Arc<BackfillRPC>,
+    metrics: Arc<BackfillerMetricsConfig>,
+    backfiller_wait_period_sec: u64,
+    rx: Receiver<()>,
+) -> Result<(), JoinError> {
+    info!("Running slot fetcher...");
+
+    if let Err(e) = backfiller_clone
+        .run_perpetual_slot_collection(
+            metrics,
+            Duration::from_secs(backfiller_wait_period_sec),
+            rpc_backfiller_clone,
+            rx,
+        )
+        .await
+    {
+        error!("Error while running perpetual slot fetcher: {}", e);
+    }
+
+    info!("Slot fetcher finished working");
+
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)]
+pub async fn run_perpetual_slot_processing<SG, BC, BP>(
```

Comment:
Do we really need a function that just calls another function with some logs?

Reply:
In general, the approach of this refactoring is to wrap all function calls made through spawn into separate functions to reduce the amount of code in the main file. I couldn't find a more elegant approach. If you have any suggestions, you are welcome.

Reply:
The odd thing is, in order to reduce the amount of code in the main fail we bloated other spaces with meaningless functions. If I were you, I'd leave them as is. So it's up to you.

Reply:
However, the primary goal of this task was to reduce the amount of code specifically in the main file and improve navigation within it.